source-mysql/datatypes.go

```diff
@@ -0,0 +1,556 @@
+		return val, nil
+	}
+	return val, fmt.Errorf("internal error: failed to encode column of type %q as backfill key", t.Type)
+}
+
+func decodeKeyFDB(t tuple.TupleElement) (any, error) {
+	switch v := t.(type) {
+	case []byte:
+		return string(v), nil
+	}
+	return t, nil
+}
+
+func transcodeTinyintUnsigned(buf []byte, val any) ([]byte, error) {
+	if sval, ok := val.(int8); ok {
+		return json.Append(buf, uint8(sval), 0)
+	}
+	return json.Append(buf, val, 0)
+}
+
+func transcodeTinyint(buf []byte, val any) ([]byte, error) {
+	return json.Append(buf, val, 0)
+}
+
+func transcodeSmallintUnsigned(buf []byte, val any) ([]byte, error) {
+	if sval, ok := val.(int16); ok {
+		return json.Append(buf, uint16(sval), 0)
+	}
+	return json.Append(buf, val, 0)
+}
+
+func transcodeSmallint(buf []byte, val any) ([]byte, error) {
+	return json.Append(buf, val, 0)
+}
+
+func transcodeMediumintUnsigned(buf []byte, val any) ([]byte, error) {
+	if sval, ok := val.(int32); ok {
+		// A MySQL 'MEDIUMINT' is a 24-bit integer value which is stored into an int32 by the client library,
+		// so we convert to a uint32 and mask off any sign-extended upper bits.
+		return json.Append(buf, uint32(sval)&0x00FFFFFF, 0)
+	}
+	return json.Append(buf, val, 0)
+}
+
+func transcodeMediumint(buf []byte, val any) ([]byte, error) {
+	return json.Append(buf, val, 0)
+}
+
+func transcodeIntUnsigned(buf []byte, val any) ([]byte, error) {
+	if sval, ok := val.(int32); ok {
+		return json.Append(buf, uint32(sval), 0)
+	}
+	return json.Append(buf, val, 0)
+}
+
+func transcodeInt(buf []byte, val any) ([]byte, error) {
+	return json.Append(buf, val, 0)
+}
+
+func transcodeBigintUnsigned(buf []byte, val any) ([]byte, error) {
+	if sval, ok := val.(int64); ok {
+		return json.Append(buf, uint64(sval), 0)
+	}
+	return json.Append(buf, val, 0)
+}
+
+func transcodeBigint(buf []byte, val any) ([]byte, error) {
+	return json.Append(buf, val, 0)
+}
+
+type mysqlTextTranscoderBackfill struct{}
+
+func (t *mysqlTextTranscoderBackfill) TranscodeJSON(buf []byte, val any) ([]byte, error) {
+	if str, ok := val.(string); ok {
+		return json.AppendEscape(buf, str, 0), nil
+	} else if bs, ok := val.([]byte); ok {
+		// Backfills always return string results as UTF-8 so no decoding is required,
+		// however we can avoid an expensive []byte -> string conversion using unsafe.
+		var str = *(*string)(unsafe.Pointer(&bs))
+		return json.AppendEscape(buf, str, 0), nil
+	} else if val == nil {
+		return append(buf, `null`...), nil
+	}
+	return nil, fmt.Errorf("internal error: text column value must be bytes or nil: got %v", val)
+}
+
+type mysqlTextTranscoderReplication struct {
+	Charset string
+}
+
+func (t *mysqlTextTranscoderReplication) TranscodeJSON(buf []byte, val any) ([]byte, error) {
+	if str, ok := val.(string); ok {
+		return json.AppendEscape(buf, str, 0), nil
+	} else if bs, ok := val.([]byte); ok {
+		// TODO(wgd): Figure out how exactly to optimize this. The UTF-8 case could
```

Comment:
Just noting this TODO, which is probably not something intended to be addressed in this PR, but rather a later thing I imagine.